frame/two-phase-election-provider/src/benchmarking.rs

```diff
@@ -0,0 +1,358 @@
+	crate::voter_at_fn!(let voter_at, all_voters, T);
+	crate::target_at_fn!(let target_at, targets, T);
+	crate::target_index_fn!(let target_index, targets, T);
+
+	let assignments = active_voters
+		.iter()
+		.map(|(voter, _stake, votes)| {
+			let percent_per_edge: InnerOf<CompactAccuracyOf<T>> =
+				(100 / votes.len()).try_into().unwrap();
+			Assignment {
+				who: voter.clone(),
+				distribution: votes
+					.iter()
+					.map(|t| (t.clone(), <CompactAccuracyOf<T>>::from_percent(percent_per_edge)))
+					.collect::<Vec<_>>(),
+			}
+		})
+		.collect::<Vec<_>>();
+
+	let compact =
+		<CompactOf<T>>::from_assignment(assignments, &voter_index, &target_index).unwrap();
+	let score = compact
+		.clone()
+		.score(&winners, stake_of, voter_at, target_at)
+		.unwrap();
+	let round = <TwoPhase<T>>::round();
+	RawSolution { compact, score, round }
+}
+
+benchmarks! {
+	where_clause {
+		where ExtendedBalance: From<InnerOf<CompactAccuracyOf<T>>>,
+		<InnerOf<CompactAccuracyOf<T>> as sp_std::convert::TryFrom<usize>>::Error: sp_std::fmt::Debug,
+		ExtendedBalance: From<InnerOf<OnChainAccuracyOf<T>>>,
+	}
+
+	on_initialize_nothing {
+		assert!(<TwoPhase<T>>::current_phase().is_off());
+	}: {
+		<TwoPhase<T>>::on_initialize(1u32.into());
+	} verify {
+		assert!(<TwoPhase<T>>::current_phase().is_off());
+	}
+
+	on_initialize_open_signed {
+		// NOTE: this benchmark currently doesn't have any components because the length of a db
+		// read/write is not captured. Otherwise, it is quite influenced by how much data
+		// `T::ElectionDataProvider` is reading and passing on.
+		assert!(<TwoPhase<T>>::snapshot().is_none());
+		assert!(<TwoPhase<T>>::current_phase().is_off());
+	}: {
+		<TwoPhase<T>>::on_initialize_open_signed();
+	} verify {
+		assert!(<TwoPhase<T>>::snapshot().is_some());
+		assert!(<TwoPhase<T>>::current_phase().is_signed());
+	}
+
+	on_initialize_open_unsigned {
+		assert!(<TwoPhase<T>>::snapshot().is_none());
+		assert!(<TwoPhase<T>>::current_phase().is_off());
+	}: {
+		<TwoPhase<T>>::on_initialize_open_unsigned(Phase::Off, 1u32.into());
+	} verify {
+		assert!(<TwoPhase<T>>::snapshot().is_some());
+		assert!(<TwoPhase<T>>::current_phase().is_unsigned());
+	}
+
+	finalize_signed_phase_accept_solution {
+		let receiver = account("receiver", 0, SEED);
+		let initial_balance = T::Currency::minimum_balance() * 10u32.into();
+		T::Currency::make_free_balance_be(&receiver, initial_balance);
+		let ready: ReadySolution<T::AccountId> = Default::default();
+		let deposit: BalanceOf<T> = 10u32.into();
+		let reward: BalanceOf<T> = 20u32.into();
+
+		assert_ok!(T::Currency::reserve(&receiver, deposit));
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance - 10u32.into());
+	}: {
+		<TwoPhase<T>>::finalize_signed_phase_accept_solution(ready, &receiver, deposit, reward)
+	} verify {
+		assert_eq!(T::Currency::free_balance(&receiver), initial_balance + 20u32.into());
+		assert_eq!(T::Currency::reserved_balance(&receiver), 0u32.into());
```

Comment:
```suggestion
		assert!(T::Currency::reserved_balance(&receiver).is_zero());
```

Feel free to ignore these nits, just things that come into my head when i scroll through lines